peekingduck/pipeline/nodes/model/mask_rcnnv1/mask_rcnn_files/roi_heads.py

```diff
@@ -0,0 +1,389 @@
+    # pylint: disable=invalid-name
+    """
+    From the results of the CNN, post process the masks
+    by taking the mask corresponding to the class with max
+    probability (which are of fixed size and directly output
+    by the CNN) and return the masks in the mask field of the BoxList.
+
+    Args:
+        x (Tensor): the mask logits
+        labels (list[BoxList]): bounding boxes that are used as
+            reference, one for ech image
+
+    Returns:
+        results (list[BoxList]): one BoxList for each image, containing
+            the extra field mask
+    """
+    mask_prob = x.sigmoid()
+
+    # select masks corresponding to the predicted classes
+    num_masks = x.shape[0]
+    boxes_per_image = [label.shape[0] for label in labels]
+    labels_ = torch.cat(labels)
+    index = torch.arange(num_masks, device=labels_.device)
+    mask_prob = mask_prob[index, labels_][:, None]
+    mask_prob_out = mask_prob.split(boxes_per_image, dim=0)
+
+    return mask_prob_out
+
+
+def expand_boxes(boxes, scale):
+    # type: (Tensor, float) -> Tensor
+    """Enlarge boxes with the specified scale"""
+    w_half = (boxes[:, 2] - boxes[:, 0]) * 0.5
+    h_half = (boxes[:, 3] - boxes[:, 1]) * 0.5
+    x_c = (boxes[:, 2] + boxes[:, 0]) * 0.5
+    y_c = (boxes[:, 3] + boxes[:, 1]) * 0.5
+
+    w_half *= scale
+    h_half *= scale
+
+    boxes_exp = torch.zeros_like(boxes)
+    boxes_exp[:, 0] = x_c - w_half
+    boxes_exp[:, 2] = x_c + w_half
+    boxes_exp[:, 1] = y_c - h_half
+    boxes_exp[:, 3] = y_c + h_half
+    return boxes_exp
+
+
+def expand_masks(mask, padding):
+    # type: (Tensor, int) -> Tuple[Tensor, float]
+    # pylint: disable=invalid-name
+    """Enlarge masks with the specified padding and returns the resultant scale"""
+    M = mask.shape[-1]
+    scale = float(M + 2 * padding) / M
+    padded_mask = F.pad(mask, (padding,) * 4)
+    return padded_mask, scale
+
+
+def paste_mask_in_image(mask, box, im_h, im_w):
+    # type: (Tensor, Tensor, int, int) -> Tensor
+    # pylint: disable=missing-function-docstring,invalid-name
+    to_remove = 1
+    w = int(box[2] - box[0] + to_remove)
+    h = int(box[3] - box[1] + to_remove)
+    w = max(w, 1)
+    h = max(h, 1)
+
+    # Set shape to [batchxCxHxW]
+    mask = mask.expand((1, 1, -1, -1))
+
+    # Resize mask
+    mask = F.interpolate(mask, size=(h, w), mode="bilinear", align_corners=False)
+    mask = mask[0][0]
+
+    im_mask = torch.zeros((im_h, im_w), dtype=mask.dtype, device=mask.device)
+    x_0 = max(box[0], 0)  # type: ignore[call-overload]
+    x_1 = min(box[2] + 1, im_w)  # type: ignore[call-overload]
+    y_0 = max(box[1], 0)  # type: ignore[call-overload]
+    y_1 = min(box[3] + 1, im_h)  # type: ignore[call-overload]
+
+    im_mask[y_0:y_1, x_0:x_1] = mask[
+        (y_0 - box[1]) : (y_1 - box[1]), (x_0 - box[0]) : (x_1 - box[0])
+    ]
+    return im_mask
+
+
+def paste_masks_in_image(masks, boxes, img_shape, padding=1):
```

Comment:
Type hint directly instead of using comments

Reply:
amended